### ISSUE TITLE
template parameter with expression containing boolean && fails to parse

### ISSUE THREAD

**Issue by ton** [swig/swig#2228]
The following valid C++ code results in a syntax error being issued by SWIG:

```
#include <type_traits>

class Value
{
  template<typename T>
  Value(typename std::enable_if<std::is_pod<T>::value && true>::type * = nullptr)
  {
  }
};
```

The problem here is boolean expression, removing `&& true`  and SWIG no longer complains. Also, wrapping the boolean expression in parentheses makes SWIG happy:

```
#include <type_traits>

class Value
{
  template<typename T>
  Value(typename std::enable_if<(std::is_pod<T>::value && true)>::type * = nullptr)
  {
  }
};
```

It would be nice for SWIG to accept this code without having to add the parentheses.

**Comment by ojwb** [comment]
It's specifically the `&&` - change that to `||` or `and` and it parses.

If you can change the header, using `and` instead of `&&` provides a workaround.

**Comment by ojwb** [comment]
Actually, wrong about `and` as that just generates token `LAND` - looks like I had the extra parentheses still in when I tested that - the extra parentheses are (as you said) a workaround at least.

What's interesting is that `LOR` (which is the token for `||`) only appears once in the grammar and there's a matching rule for `LAND` (token for `&&`) there.  `LAND` appears in various other places for C++11 rvalue references. but that suggests to me that one of the shift-reduce conflicts in the grammar is for this case, and it's resolved the way we don't want here (but probably the way we do want for other cases).  I suspect this will be tricky to fix.

**Comment by ton** [comment]
I see. That may explain why in some cases I've seen `&&` in template parameters where SWIG does not complain about it. I'll try to come up with a test case for those as well, that might help.

**Comment by ton** [comment]
The following illustrates a case where parentheses around the boolean expression are not needed:

```
class Value
{
  template<typename T>
  Value(typename std::enable_if<std::is_pod<T>::value && std::integral_constant<bool, true>::value>::type * = nullptr)
  {
  }
};
```

**Comment by ojwb** [comment]
I found a way to fix this via parser error recovery, but unfortunately it introduces other issues so needs more work.  Hopefully I can sort those out but I've backed it out for now.

> The following illustrates a case where parentheses around the boolean expression are not needed for SWIG to not error out on the code:

I think the difference there is that `(std::is_pod<T>::value && true)` looks like a function argument list (if you ignore that `true` is a token rather than an identifier at least).

**Comment by ojwb** [comment]
Simply telling BISON to generate a GLR parser is enough to fix this.  Patch to do so:

```
diff --git a/Source/CParse/parser.y b/Source/CParse/parser.y
index 702cbfbf..296bd26f 100644
--- a/Source/CParse/parser.y
+++ b/Source/CParse/parser.y
@@ -15,6 +15,8 @@
 
 %require "3.5"
 
+%glr-parser
+
 /* There are a small number of known shift-reduce conflicts in this file, fail
    compilation if any more are introduced.
 
```

This works because the parser then evaluates both the shift and reduce alternatives from the conflict in parallel rather than always resolving by shifting.  Always shifting fails if (as in this example) reducing was actually the right answer.

**Comment by ojwb** [comment]
I think we should consider enabling `%glr-parser` as it should make it much simpler to solve this and other problems we've had to punt on due to them introducing more conflicts.  It will probably also allow us to eliminate hacks we have previously added to try to address awkward parsing cases.

If we're going to do this, doing it early in 4.3.0 development seems appropriate.  I don't really expect much fall-out, but maximising the time for us to uncover any problems seems prudent.

So adding to the swig-4.3 milestone.

**Comment by wsfulton** [comment]
Yes, this seems to be a good strategy.

**Comment by ojwb** [comment]
Unfortunately just switching to a GLR parser breaks some testcases so I've had to revert that for now.

What I think must be happening is due to currently any shift/reduce conflicts getting resolved by shifting, whereas with a GLR parser both parses get tried and so I suspect somehow the reduce one gets used instead.  Or something like that.  I'll try to get to the bottom of this.

I thought I had run the full test suite before with `%glr-parser` added, but maybe I only did for one target language.

**Comment by ojwb** [comment]
I need to dig into what's going wrong when we generate a GLR parser and this isn't realistically going to happen before 4.3.0 now.

**Comment by wsfulton** [comment]
Another expression patch instead of the GLR parser - I defer to @ojwb here and have bumped to swig-4.6 for now.

**Comment by ojwb** [comment]
I find it hard to tell what #3432 is actually changing, but it seems to have failed a lot of CI jobs and was closed by the submitter so I'm guessing it's not a good approach.

Switching to a GLR parser fixes the case here, but unfortunately causes other problems and I haven't found the time to dig into what's going on there.

If someone had a clean working patch for this I wouldn't block applying it, but probably poking at SWIG with a GLR parser is more productive as it should help make a lot of parsing problems easier to address.
